internal/operator-controller/authorization/rbac.go

```diff
@@ -0,0 +1,605 @@
+			rules = append(rules, manifestClusterRole.Rules...)
+		}
+	case "Role":
+		if manifestRole, ok := ec.extraRoles[types.NamespacedName{Namespace: roleBinding.Namespace, Name: roleBinding.RoleRef.Name}]; ok {
+			rules = append(rules, manifestRole.Rules...)
+		}
+	}
+
+	if err := validation.ConfirmNoEscalation(ctx, ec.ruleResolver, rules); err != nil {
+		return err
+	}
+	return nil
+}
+
+var fullAuthority = []rbacv1.PolicyRule{
+	{Verbs: []string{"*"}, APIGroups: []string{"*"}, Resources: []string{"*"}},
+	{Verbs: []string{"*"}, NonResourceURLs: []string{"*"}},
+}
+
+// TODO: Investigate replacing this regex parsing with structured error handling once there are
+//
+//	structured RBAC errors introduced by https://github.com/kubernetes/kubernetes/pull/130955.
+//
+// parseEscalationErrorForMissingRules attempts to extract specific RBAC permissions
+// that were denied due to escalation prevention from a given error's text.
+// It returns the list of extracted PolicyRules and an error.
+// Note: If parsing is successful, the returned error is derived from the *input* error's
+// message (specifically the part indicating the escalation attempt), not an error
+// encountered during the parsing process itself. If parsing fails due to an unexpected
+// error format, a distinct parsing error is returned.
+func parseEscalationErrorForMissingRules(ecError error) ([]rbacv1.PolicyRule, error) {
+	// errRegex captures the standard prefix of an escalation error message
+	errRegex := regexp.MustCompile(`(?s)^(user ".*" \(groups=.*\) is attempting to grant RBAC permissions not currently held):.*?$`)
+	// permRegex extracts the details (APIGroups, Resources, Verbs) of individual permissions listed within the error message
+	permRegex := regexp.MustCompile(`{APIGroups:\[("[^"]*")], Resources:\[("[^"]*")], Verbs:\[("[^"]*")]}`)
+
+	errMatches := errRegex.FindAllStringSubmatch(ecError.Error(), -1)
+	// Check if the main error message prefix was matched and captured
+	if len(errMatches) == 0 || len(errMatches[0]) < 2 {
+		// The error format doesn't match the expected pattern for escalation errors
+		return nil, fmt.Errorf("failed to parse escalation error: unexpected format: %w", ecError)
+	}
+
+	// Extract permissions using permRegex
+	permissions := []rbacv1.PolicyRule{}
+	permMatches := permRegex.FindAllStringSubmatch(ecError.Error(), -1)
+	for _, match := range permMatches {
+		// Ensure the match has the expected number of capture groups
+		if len(match) < 4 {
+			continue // Skip malformed permission strings
+		}
+		permissions = append(permissions, rbacv1.PolicyRule{
+			APIGroups: []string{strings.Trim(match[1], `"`)},
+			Resources: []string{strings.Trim(match[2], `"`)},
+			Verbs:     []string{strings.Trim(match[3], `"`)},
+		})
+	}
+
+	// Return the extracted permissions and the captured escalation message prefix as the error context
+	return permissions, errors.New(errMatches[0][1])
```

Comment:
I think we should not mix the concerns of the returned error since they are semantically different.

If we produce errors in this function trying to parse the message, those are parsing errors. If parsing is successful, we may have parsed some "resolution errors" out of the string. The parsing error and the resolution error should probably be returned as two separate values. Maybe we should return `[]rbacv1.PolicyRule, string, error` where the string is the resolution errors we parse out of the regex, and the error is reserved for "we couldn't parse this error message string".